midend/predication.cpp

```diff
@@ -13,11 +13,120 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
-
 #include "predication.h"
 #include "frontends/p4/cloner.h"
 
+
 namespace P4 {
+/// convert an expression into a string that uniqely identifies the value referenced
+/// return null cstring if not a reference to a constant thing.
+static cstring expr_name(const IR::Expression *exp) {
+    if (auto p = exp->to<IR::PathExpression>())
+        return p->path->name;
+    if (auto m = exp->to<IR::Member>()) {
+        if (auto base = expr_name(m->expr))
+            return base + "." + m->member;
+    } else if (auto a = exp->to<IR::ArrayIndex>()) {
+        if (auto k = a->right->to<IR::Constant>())
+            if (auto base = expr_name(a->left))
+                return base + "." + std::to_string(k->asInt()); }
+    return cstring();
+}
+
+bool Predication::EmptyStatementRemover::preorder(IR::BlockStatement * block) {
```

Comment:
There are some other passes which do this as well, e.g., DoSimplifyControlFlow.

Reply:
_DoSimplifyControlFlow_, does more than removing _EmptyStatement_, so it's changing other code in the Predication pass.